drivers/mdio/mdio_dwcxgmac.c

```diff
@@ -0,0 +1,159 @@
+
+	ioaddr = (mem_addr_t)DEVICE_MMIO_GET(dev);
+
+	retval = mdio_busy_wait((ioaddr + CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_OFST),
+				CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SBUSY_SET_MSK);
+	if (retval) {
+		LOG_ERR("MDIO device busy wait timedout");
+		return retval;
+	}
+
+	(void)k_mutex_lock(&data->mdio_transfer_lock, K_FOREVER);
+	/* Set port as Clause 22 */
+	reg_data = sys_read32(ioaddr + CORE_MDIO_CLAUSE_22_PORT_OFST);
+	reg_data |= BIT(prtad);
+	sys_write32(reg_data, ioaddr + CORE_MDIO_CLAUSE_22_PORT_OFST);
+
+	mdio_addr = CORE_MDIO_SINGLE_COMMAND_ADDRESS_RA_SET(devad) |
+		    CORE_MDIO_SINGLE_COMMAND_ADDRESS_PA_SET(prtad);
+	sys_write32(mdio_addr, (ioaddr + CORE_MDIO_SINGLE_COMMAND_ADDRESS_OFST));
+
+	mdio_data = CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SDATA_SET(data_in) |
+		    CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CMD_SET(rw) |
+		    CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CR_SET(cfg->clk_range) |
+		    CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_CRS_SET(cfg->clk_range_sel) |
+		    CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SBUSY_SET(1u);
+
+	sys_write32(mdio_data, (ioaddr + CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_OFST));
+
+	retval = mdio_busy_wait((ioaddr + CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_OFST),
+				CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SBUSY_SET_MSK);
+	if (retval) {
+		LOG_ERR("transfer timedout");
+	} else {
+		if (data_out) {
+			*data_out = CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_SDATA_GET(
+				sys_read32(ioaddr + CORE_MDIO_SINGLE_COMMAND_CONTROL_DATA_OFST));
+		}
+	}
+
+	(void)k_mutex_unlock(&data->mdio_transfer_lock);
+
+	return retval;
+}
+
+static int mdio_dwcxgmac_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+			      uint16_t *data)
+{
+	return mdio_transfer(dev, prtad, devad, 1, 0, data);
+}
+
+static int mdio_dwcxgmac_write(const struct device *dev, uint8_t prtad, uint8_t devad,
+			       uint16_t data)
+{
+	return mdio_transfer(dev, prtad, devad, 0, data, NULL);
+}
+
+static void mdio_dwcxgmac_bus_enable(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
+
+static void mdio_dwcxgmac_bus_disable(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
+
+static int mdio_dwcxgmac_initialize(const struct device *dev)
+{
+	struct mdio_dwcxgmac_dev_data *const data = (struct mdio_dwcxgmac_dev_data *)dev->data;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+	k_mutex_init(&data->mdio_transfer_lock);
+	return 0;
+}
+
+static const struct mdio_driver_api mdio_dwcxgmac_driver_api = {
+	.read = mdio_dwcxgmac_read,
+	.write = mdio_dwcxgmac_write,
+	.bus_enable = mdio_dwcxgmac_bus_enable,
+	.bus_disable = mdio_dwcxgmac_bus_disable,
+};
+
+#define MDIO_DWCXGMAC_CONFIG(n)                                                                    \
+	static const struct mdio_dwcxgmac_dev_config mdio_dwcxgmac_dev_config_##n = {              \
+		DEVICE_MMIO_ROM_INIT(DT_DRV_INST(n)),                                              \
+		.clk_range = DT_INST_PROP(n, csr_clock_indx),                                      \
```

Comment:
do you happen to know what CSR stands for? seems every vendor has an mdio peripheral with this acronym but I have not been able to deduce or find the unabbreviated term

Reply:
Here the CSR stands for configuration and Status Register

Reply:
Here CSR stands for Control and Status Register.

Reply:
thanks, that's probably all it is I guess on the other chips too